### ISSUE TITLE
GNN edge features don't get updated

### ISSUE THREAD

**Issue by kks32** [Centrum-IntelliPhysics/GNS_vs_NOs#1]
There is a bug in your GNN edge feature update:

For e.g., https://github.com/Centrum-IntelliPhysics/GNS_vs_NOs/blob/main/2D_nonlinear_SWE/models_gns.py#L60 doesn't update as you expect. 

The `update()` method correctly uses `aggr_out` to update the node features, but it then returns `return x_updated, edge_attr`. The `edge_attr` it returns is the original, unchanged edge attributes from the beginning of the layer's forward pass.

The `propagate()` call returns what `update()` returns. So, in your forward method, the variable `edge_attr_updated` actually receives the original edge attributes.

Here's a code that demonstrates this issue and a possible fix [This is not an optimized fix]

```
import torch
import torch.nn as nn
import torch_geometric.nn as pyg_nn

def build_mlp(input_dim, hidden_dims, output_dim):
    layers = []
    last_dim = input_dim
    for h in hidden_dims:
        layers.append(nn.Linear(last_dim, h))
        layers.append(nn.GELU())
        last_dim = h
    layers.append(nn.Linear(last_dim, output_dim))
    return layers

# ============= BUGGY VERSION =============
class GNSLayer(pyg_nn.MessagePassing):
    def __init__(self, node_dim, edge_dim, hidden_dim, n_mlp_layers=2):
        super().__init__(aggr='mean')
        
        self.edge_mlp = nn.Sequential(
            *build_mlp(2 * node_dim + edge_dim,
                      [hidden_dim] * n_mlp_layers,
                      edge_dim),
            nn.LayerNorm(edge_dim)
        )
        
        self.node_mlp = nn.Sequential(
            *build_mlp(node_dim + edge_dim,
                      [hidden_dim] * n_mlp_layers,
                      node_dim),
            nn.LayerNorm(node_dim)
        )
        
    def forward(self, x, edge_index, edge_attr):
        x_res = x
        edge_res = edge_attr
        x_updated, edge_attr_updated = self.propagate(edge_index, x=x, edge_attr=edge_attr)
        return x_res + x_updated, edge_res + edge_attr_updated

    def message(self, x_i, x_j, edge_attr):
        edge_input = torch.cat([x_i, x_j, edge_attr], dim=-1)
        edge_attr_updated = self.edge_mlp(edge_input)
        return edge_attr_updated

    def update(self, aggr_out, x, edge_attr):
        node_input = torch.cat([x, aggr_out], dim=-1)
        x_updated = self.node_mlp(node_input)
        return x_updated, edge_attr  # BUG: returns ORIGINAL edge_attr


# ============= FIXED VERSION =============
class GNSLayerFixed(pyg_nn.MessagePassing):
    def __init__(self, node_dim, edge_dim, hidden_dim, n_mlp_layers=2):
        super().__init__(aggr='mean')
        
        self.edge_mlp = nn.Sequential(
            *build_mlp(2 * node_dim + edge_dim,
                      [hidden_dim] * n_mlp_layers,
                      edge_dim),
            nn.LayerNorm(edge_dim)
        )
        
        self.node_mlp = nn.Sequential(
            *build_mlp(node_dim + edge_dim,
                      [hidden_dim] * n_mlp_layers,
                      node_dim),
            nn.LayerNorm(node_dim)
        )
        
    def forward(self, x, edge_index, edge_attr):
        x_res = x
        edge_res = edge_attr
        
        # Store updated edge attributes during message passing
        self.updated_edge_attr = None
        x_updated = self.propagate(edge_index, x=x, edge_attr=edge_attr)
        edge_attr_updated = self.updated_edge_attr
        
        return x_res + x_updated, edge_res + edge_attr_updated

    def message(self, x_i, x_j, edge_attr):
        edge_input = torch.cat([x_i, x_j, edge_attr], dim=-1)
        edge_attr_updated = self.edge_mlp(edge_input)
        
        # Store for later use
        self.updated_edge_attr = edge_attr_updated
        
        return edge_attr_updated

    def update(self, aggr_out, x):
        node_input = torch.cat([x, aggr_out], dim=-1)
        x_updated = self.node_mlp(node_input)
        return x_updated


# Demonstrate bug
torch.manual_seed(42)

edge_index = torch.tensor([[0, 1],
                           [1, 0]], dtype=torch.long)

x = torch.tensor([[1.0, 1.0], 
                  [2.0, 2.0]], dtype=torch.float)

edge_attr_buggy = torch.tensor([[1.0, 1.0], 
                                [2.0, 2.0]], dtype=torch.float)

edge_attr_fixed = edge_attr_buggy.clone()

# Create layers with SAME weights for fair comparison
buggy_layer_1 = GNSLayer(node_dim=2, edge_dim=2, hidden_dim=4)
buggy_layer_2 = GNSLayer(node_dim=2, edge_dim=2, hidden_dim=4)
buggy_layer_3 = GNSLayer(node_dim=2, edge_dim=2, hidden_dim=4)

fixed_layer_1 = GNSLayerFixed(node_dim=2, edge_dim=2, hidden_dim=4)
fixed_layer_2 = GNSLayerFixed(node_dim=2, edge_dim=2, hidden_dim=4)
fixed_layer_3 = GNSLayerFixed(node_dim=2, edge_dim=2, hidden_dim=4)

# Copy weights to make them identical
fixed_layer_1.load_state_dict(buggy_layer_1.state_dict())
fixed_layer_2.load_state_dict(buggy_layer_2.state_dict())
fixed_layer_3.load_state_dict(buggy_layer_3.state_dict())

print(f"\nInitial edge_attr: {edge_attr_buggy[0].tolist()}, {edge_attr_buggy[1].tolist()}")

# ===== BUGGY VERSION =====
print("BUGGY VERSION (edges just double):")


x_buggy = x.clone()
x_buggy, edge_attr_buggy = buggy_layer_1(x_buggy, edge_index, edge_attr_buggy)
print(f"After Layer 1: {edge_attr_buggy[0].tolist()}, {edge_attr_buggy[1].tolist()}")

x_buggy, edge_attr_buggy = buggy_layer_2(x_buggy, edge_index, edge_attr_buggy)
print(f"After Layer 2: {edge_attr_buggy[0].tolist()}, {edge_attr_buggy[1].tolist()}")

x_buggy, edge_attr_buggy = buggy_layer_3(x_buggy, edge_index, edge_attr_buggy)
print(f"After Layer 3: {edge_attr_buggy[0].tolist()}, {edge_attr_buggy[1].tolist()}")

# ===== FIXED VERSION =====
print("FIXED VERSION (edges learn from edge_mlp):")

x_fixed = x.clone()
x_fixed, edge_attr_fixed = fixed_layer_1(x_fixed, edge_index, edge_attr_fixed)
print(f"After Layer 1: {edge_attr_fixed[0].tolist()}, {edge_attr_fixed[1].tolist()}")

x_fixed, edge_attr_fixed = fixed_layer_2(x_fixed, edge_index, edge_attr_fixed)
print(f"After Layer 2: {edge_attr_fixed[0].tolist()}, {edge_attr_fixed[1].tolist()}")

x_fixed, edge_attr_fixed = fixed_layer_3(x_fixed, edge_index, edge_attr_fixed)
print(f"After Layer 3: {edge_attr_fixed[0].tolist()}, {edge_attr_fixed[1].tolist()}")
```
> Result :
Initial edge_attr: [1.0, 1.0], [2.0, 2.0]
BUGGY VERSION (edges just double):
After Layer 1: [2.0, 2.0], [4.0, 4.0]
After Layer 2: [4.0, 4.0], [8.0, 8.0]
After Layer 3: [8.0, 8.0], [16.0, 16.0]
FIXED VERSION (edges learn from edge_mlp):
After Layer 1: [1.9993153810501099, 0.0006846189498901367], [2.9958081245422363, 1.0041918754577637]
After Layer 2: [2.9990434646606445, -0.9990434646606445], [3.995574951171875, 0.004425227642059326]
After Layer 3: [2.000222682952881, -0.00022262334823608398], [2.9972803592681885, 1.002719759941101]

**Comment by dnayak2-jh** [comment]
Hello,

Thank you for your comment. While a typical GNS implementation updates both node and edge features, in this particular work we use structured grids that do not deform with time. The primary quantities of interest exist as nodal fields only. Therefore, in the current implementation, edge features serve as static conditioning information: they inform the architecture about the spatial relationships and positions but do not themselves get updated over time.

Having said that, we do believe that incorporating dynamic edge feature updates could provide enhancements to the accuracy of the GNS surrogate at inference. We will certainly explore this modification and add it to our model_gns.py file.

Thank you so much again for the constructive feedback and taking the time to explore our repository!
